### PR TITLE
implement the prerelease Linux build

### PR DESCRIPTION
**Pull request by oconnor663**
This build includes KBFS and Electron. It's a mode, just like "prod"/"staging"/"devel". (As far as Go tags are concerned, it sets both `production` and `prerelease`.) Instead of pushing to the server-ops repo, it pushes to S3, [alongside our OSX prerelease packages](https://s3.amazonaws.com/prerelease.keybase.io/index.html). Prerelease packages have the same name as the production version ("keybase"), so are incompatible with production. Installing a prerelease Linux package will point your package manager to our S3-backed repo instead of our normal server-ops-backed repo for updates, so you will remain on the prerelease track.

As part of all these changes, I've refactored the build scripts to only build binaries once (this saves time, but more importantly it records a single timestamp-based prerelease version number). I've also moved all of our build scripts out of server-ops and into client. The `client/packaging/release/release.sh` script now runs `client/packaging/linux/docker_build.sh` for Linux.

I will test all the update paths by hand with VMs before I land this. In particular, the staging packages will now live in a separate server-side package repository, instead of living in the same repository as production, so URLs used in postinstall steps have changed and need to be verified. This will break updates for anyone currently using the staging packages. (Is there anyone in the wild using these?)

@cjb @patrickxb @maxtaco

### CODE REVIEW

**Comment by cjb** [comment]
:thumbsup: Looks great, thanks!  Adding @gabriel too.

**Comment by patrickxb** [comment]
Just a question (without reading any code):  can we still make a CLI-only brew/linux/windows release?

**Comment by oconnor663** [comment]
The way I've configured it now, "production" mode is unchanged, but "prerelease"/"staging"/"devel" now include KBFS and Electron. My guess (desparate hope?) is that in the future the production package will be expanded to include that stuff too (which here amounts to deleting an if-statement in build_binaries.sh). If instead we decide we want multiple production packages, that will require more plumbing, and will also make me scared :)

**Comment by patrickxb** [comment]
Ok, sounds good to me.

I believe we'd like to get rid of CLI-only production releases ASAP, but it might make sense to make a few more as bugs get fixed.

**Comment by oconnor663** [comment]
Yep, that should be simple. Assuming I didn't hose everything with these changes I mean...

File: go/updater/updater.go [hunk 1]
```diff
@@ -101,6 +101,18 @@ func (u *Updater) checkForUpdate(skipAssetDownload bool, force bool, requested b
 		return
 	}
 
+	// do only if instructions are empty/nil
+	if update.Instructions == nil || *update.Instructions == "" {
+		instructions, err := libkb.PlatformSpecificUpgradeInstructionsString()
+		u.log.Info("Updating instructions! %s, err: %s", instructions, err)
```
Comment (gabriel) [inline go/updater/updater.go]: Can we clean up the logging here make this a debug and move the err to the if err != nil part?

File: go/updater/updater.go [hunk 2]
```diff
@@ -101,6 +101,18 @@ func (u *Updater) checkForUpdate(skipAssetDownload bool, force bool, requested b
 		return
 	}
 
+	// do only if instructions are empty/nil
```
Comment (gabriel) [inline go/updater/updater.go]: This comment is kind of obvious from the code? Maybe a better comment is:

// Update instruction might be empty, if so we'll fill in with the default instructions for the platform

File: protocol/avdl/update_common.avdl [hunk 3]
```diff
@@ -24,6 +24,7 @@ protocol updateCommon {
     string version; // Sematic version, e.g. 1.2.3-400
     string name;
     string description;
+    union { null, string } instructions; // String representing distro-specific upgrade instructions
```
Comment (gabriel) [inline protocol/avdl/update_common.avdl]: I think using default empty string for empty (instead of union null) is pretty common in the protocol?

**Comment by gabriel** [comment]
lg2m but might want other feedback... my comments relate more to marco's changes and are only optional suggestions

File: packaging/linux/deb/package_binaries.sh [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+if [ "$mode" = "production" ] ; then
+  repo_url="http://dist.keybase.io/linux/deb/repo"
+elif [ "$mode" = "prerelease" ] ; then
+  repo_url="http://s3.amazonaws.com/prerelease.keybase.io/deb"
+elif [ "$mode" = "staging" ] ; then
+  # Note: This doesn't exist yet. But we need to be distinct from the
+  # production URL, because we're moving to a model where we build a clean repo
+  # every time, rather than adding to an existing one. (For S3 compatibility.)
+  repo_url="http://dist.keybase.io/linux/deb_staging/repo"
```
Comment (patrickxb) [inline packaging/linux/deb/package_binaries.sh]: fine with me to make this change.  we never made any promises not to change staging releases...

File: packaging/linux/inside_docker_main.sh [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+END
+
+  aws s3 cp --content-type application/json "$json_tmp" s3://prerelease.keybase.io/update-linux-prod.json
+}
+
+release_serverops() {
+  serverops_clone="/root/server-ops"
+  if [ "$mode" = staging ] ; then
+    deb_repo="$serverops_clone/prod/linux/deb_staging"
+    rpm_repo="$serverops_clone/prod/linux/rpm_staging"
```
Comment (patrickxb) [inline packaging/linux/inside_docker_main.sh]: These look good to me too.

**Comment by patrickxb** [comment]
LGTM.

File: protocol/avdl/update_common.avdl (hunk 3 above)
Comment (oconnor663) [inline protocol/avdl/update_common.avdl]: @MarcoPolo if you want to change this API, can I leave that for another PR?
